### ISSUE TITLE
Add paragraph‑ and delimiter‑based text chunking (in addition to word‑count) for Single Speaker

### ISSUE THREAD

**Issue by gileneusz** [Enemyx-net/VibeVoice-ComfyUI#108]
Sorry for long issue, my AI is crazy lately. There is an overview and later full description:

Overview:

Hi! Thanks for the excellent v1.2.4 update — the new word‑based chunking already improves long‑text handling. I’d like to propose two **optional** chunking modes for the Single Speaker node: **by paragraphs** and **by a custom delimiter** (e.g., “\[stop]”). This would let users align pauses with meaningful boundaries (sections, cues) for more natural prosody in long‑form scripts, while keeping today’s word‑based default untouched.
Details and examples below. *(Drafted with help from my AI assistant, so minor inaccuracies are possible.)*

Details:

**Why another chunking mode?**
For long‑form narration, word‑count splits can still break rhetorical units (sections, cues, list items), which sometimes makes prosody feel rushed or oddly paused. Two additional **optional** modes would give users more control without breaking existing workflows:

1. **Paragraph‑based chunking** (by number of paragraphs per chunk)
2. **Delimiter‑based chunking** (by a custom token like `[stop]`, `---`, `[[CHUNK]]`, etc.)

**Proposed UX / API (Single Speaker node):**

* `chunking_mode`: `words` (default, keeps current behavior) | `paragraphs` | `delimiter` | `auto`
* `max_paragraphs_per_chunk`: `int` (e.g., 1–10, default `2`)
* `custom_delimiter`: `string` (e.g., `[stop]`)
* `keep_delimiter_pause`: `bool` (default `true`) — insert a short pause where the delimiter was (don’t speak it)
* `insert_pause_ms`: `int` (0–1000, default `200`) — optional silence between concatenated chunks
* *Backwards‑compatible defaults*: `chunking_mode="words"`; existing `max_words_per_chunk` unchanged

**Examples**

**A) Paragraph mode (`max_paragraphs_per_chunk = 2`):**

```
Paragraph 1: The demo starts with a short intro about our product and its goals.

Paragraph 2: Next, we outline the core features and who benefits the most.

Paragraph 3: In the third section, we describe real-world use cases and results.

Paragraph 4: Finally, we wrap up with next steps and contact information.
```

Expected chunks:

* Chunk 1 → Paragraphs 1–2
* Chunk 2 → Paragraphs 3–4

**B) Delimiter mode (`custom_delimiter = "[stop]"`):**

```
Intro: Welcome to the audio guide for our museum. Please follow the numbered rooms.
[stop]
Room 1: On your left, you’ll see a series of early sketches by the artist.
Room 2: Ahead, the large canvas explores light and motion in urban spaces.
[stop]
Conclusion: Thank you for listening. We hope you enjoyed the tour!
```

Expected chunks:

* Chunk 1 → “Intro: …” (ends at `[stop]`)
* Chunk 2 → “Room 1 … Room 2 …” (ends at next `[stop]`)
* Chunk 3 → “Conclusion: …”

**Implementation sketch (non‑breaking) — file‑level pointers**

* `nodes/base_vibevoice.py`

  * Add helpers:

    * `_split_text_by_paragraphs(text: str, max_paras: int) -> List[str]`
      Split on blank lines: `re.split(r'\r?\n\s*\r?\n+', text)`; pack up to `max_paragraphs_per_chunk` per chunk.
    * `_split_text_by_delimiter(text: str, delim: str) -> List[str]`
      Use `re.split(re.escape(delim))`, strip empties.
    * `_auto_split_text(text: str, ...)`
      Prefer delimiter if present; else paragraph if multiple paragraphs; else fallback to existing `_split_text_into_chunks(text, max_words)`.
  * (Optional) When concatenating, insert silence between chunks if `insert_pause_ms > 0`: append `torch.zeros(int(sample_rate * insert_pause_ms/1000))` between waveforms.

* `nodes/single_speaker_node.py`

  * Extend `INPUT_TYPES` with:

    ```py
    "chunking_mode": ("COMBO", {"default": "words", "choices": ["words","paragraphs","delimiter","auto"]}),
    "max_paragraphs_per_chunk": ("INT", {"default": 2, "min": 1, "max": 10, "step": 1}),
    "custom_delimiter": ("STRING", {"default": "[stop]"}),
    "keep_delimiter_pause": ("BOOLEAN", {"default": True}),
    "insert_pause_ms": ("INT", {"default": 200, "min": 0, "max": 2000, "step": 50}),
    ```
  * Update `generate_speech(...)` signature to accept these params.
  * Before current word‑based path, select splitter:

    ```py
    if chunking_mode == "delimiter" and custom_delimiter and custom_delimiter in final_text:
        text_chunks = self._split_text_by_delimiter(final_text, custom_delimiter)
    elif chunking_mode == "paragraphs":
        text_chunks = self._split_text_by_paragraphs(final_text, max_paragraphs_per_chunk)
    elif chunking_mode == "auto":
        text_chunks = self._auto_split_text(final_text, max_words_per_chunk, max_paragraphs_per_chunk, custom_delimiter)
    else:
        text_chunks = self._split_text_into_chunks(final_text, max_words_per_chunk)  # current behavior
    ```
  * Reuse the **same** prepared `voice_samples` and `seed` across chunks (as done now) to keep timbre consistent; then concatenate waveforms, optionally inserting silence when `insert_pause_ms > 0`.

**Edge cases & Safety:**

* If a *single paragraph* still exceeds `max_words_per_chunk`, optionally re‑split it by words inside that paragraph (fallback).
* Skip empty segments from repeated delimiters.
* Trim delimiters from spoken text (`keep_delimiter_pause=True` means “pause only”).
* Keep current defaults so existing workflows stay exactly the same.

**Docs:**

* README → Single Speaker parameters table: list `chunking_mode`, `max_paragraphs_per_chunk`, `custom_delimiter`, `insert_pause_ms`, `keep_delimiter_pause`. (Current README already documents `max_words_per_chunk` after v1.2.4.) ([[GitHub](https://github.com/Enemyx-net/VibeVoice-ComfyUI)][3])

---

**Note:** This example was prepared by my AI assistant and may contain minor inaccuracies about the code layout or style. Please treat it as a high‑level proposal — happy to adjust if anything doesn’t fit the project conventions.

**Labels:** `enhancement`, `UX`, `audio-prosody`

---

### (Optional) Minimal diff‑style sketch (illustrative)

```diff
--- a/nodes/single_speaker_node.py
+++ b/nodes/single_speaker_node.py
@@ class VibeVoiceSingleSpeakerNode:
-    def INPUT_TYPES(cls):
+    def INPUT_TYPES(cls):
         return {
           "required": {
             ...
             "top_p": ("FLOAT", {...}),
-            "max_words_per_chunk": ("INT", {...}),
+            "max_words_per_chunk": ("INT", {...}),
+            "chunking_mode": ("COMBO", {"default":"words","choices":["words","paragraphs","delimiter","auto"]}),
+            "max_paragraphs_per_chunk": ("INT", {"default":2,"min":1,"max":10,"step":1}),
+            "custom_delimiter": ("STRING", {"default":"[stop]"}),
+            "keep_delimiter_pause": ("BOOLEAN", {"default": True}),
+            "insert_pause_ms": ("INT", {"default":200,"min":0,"max":2000,"step":50}),
           }
         }

-    def generate_speech(self, text: str = "", ..., max_words_per_chunk: int = 250):
+    def generate_speech(self, text: str = "", ..., max_words_per_chunk: int = 250,
+                        chunking_mode: str = "words", max_paragraphs_per_chunk: int = 2,
+                        custom_delimiter: str = "[stop]", keep_delimiter_pause: bool = True,
+                        insert_pause_ms: int = 200):
         ...
-        word_count = len(final_text.split())
-        if word_count > max_words_per_chunk:
-            text_chunks = self._split_text_into_chunks(final_text, max_words_per_chunk)
+        if chunking_mode == "delimiter" and custom_delimiter and custom_delimiter in final_text:
+            text_chunks = self._split_text_by_delimiter(final_text, custom_delimiter)
+        elif chunking_mode == "paragraphs":
+            text_chunks = self._split_text_by_paragraphs(final_text, max_paragraphs_per_chunk)
+        elif chunking_mode == "auto":
+            text_chunks = self._auto_split_text(final_text, max_words_per_chunk, max_paragraphs_per_chunk, custom_delimiter)
+        else:
+            text_chunks = self._split_text_into_chunks(final_text, max_words_per_chunk)
         ...
-        combined_waveform = torch.cat(waveforms, dim=-1)
+        if insert_pause_ms > 0:
+            sr = sample_rate
+            pad = torch.zeros(1, int(sr * insert_pause_ms / 1000.0), device=waveforms[0].device)
+            pad_chunks = []
+            for i, w in enumerate(waveforms):
+                pad_chunks.append(w)
+                if i < len(waveforms) - 1:
+                    pad_chunks.append(pad)
+            waveforms = pad_chunks
+        combined_waveform = torch.cat(waveforms, dim=-1)
```

```diff
--- a/nodes/base_vibevoice.py
+++ b/nodes/base_vibevoice.py
@@ class BaseVibeVoice:
+    def _split_text_by_paragraphs(self, text: str, max_paras: int) -> List[str]:
+        paras = [p.strip() for p in re.split(r'\r?\n\s*\r?\n+', text) if p.strip()]
+        chunks, buf = [], []
+        for p in paras:
+            buf.append(p)
+            if len(buf) >= max_paras:
+                chunks.append("\n\n".join(buf)); buf = []
+        if buf: chunks.append("\n\n".join(buf))
+        return chunks or [text]
+
+    def _split_text_by_delimiter(self, text: str, delim: str) -> List[str]:
+        parts = [s.strip() for s in re.split(re.escape(delim), text) if s.strip()]
+        return parts or [text]
+
+    def _auto_split_text(self, text: str, max_words: int, max_paras: int, delim: str) -> List[str]:
+        if delim and delim in text:
+            return self._split_text_by_delimiter(text, delim)
+        if re.search(r'\r?\n\s*\r?\n+', text):
+            return self._split_text_by_paragraphs(text, max_paras)
+        return self._split_text_into_chunks(text, max_words)  # existing v1.2.4 logic
```
